packages/compose/src/hooks/use-instance-id/index.js

```diff
@@ -0,0 +1,48 @@
+/**
+ * WordPress dependencies
+ */
+import { useMemo, useEffect } from '@wordpress/element';
+
+/**
+ * Array to keep track of allocated ids.
+ */
+const allocatedIds = [];
+
+/**
+ * Find an unallocated id.
+ */
+function findId() {
+	let id = allocatedIds.findIndex( ( allocated ) => ! allocated );
```

Comment:
This is slightly less performant than just increment the id, any concern about that?

Reply:
I'm not concerned as this only executes when a component mounts. Incrementing ids is simple, but it has no end. The highest number possible here is the amount of mounted components per page.

Reply:
`withInstanceId` has counters for _each_ component. I don't think we can do that with hooks. The counter would be per page load. So the number could potentially reach much higher than before. That's why I decided to reuse ids for unmounted components.

Reply:
Use a clean up stack:

```js
const nextId = 0;
const freedIds = [];
function findId() {
	if ( freedIds.length ) {
		return freedIds.pop();
	}
	return nextId++;
}
function freeId( id ) {
	freedIds.push( id );
}
```

Reply:
That works too. Either you have a pool of allocated ids or a pool of free ids.

Reply:
The point of the stack is to avoid the linear lookup.

Reply:
Sounds good, I’ll adjust it. Thanks!